lib/web_ui/dev/generate_fallback_font_data.dart

```diff
@@ -0,0 +1,269 @@
+          'Google Fonts',
+    );
+  }
+
+  @override
+  final String name = 'generate-fallback-font-data';
+
+  @override
+  final String description = 'Generate fallback font data from GoogleFonts';
+
+  String get apiKey => stringArg('key');
+
+  @override
+  Future<bool> run() async {
+    await _generateFallbackFontData();
+    return true;
+  }
+
+  Future<void> _generateFallbackFontData() async {
+    if (apiKey.isEmpty) {
+      throw UsageException('No Google Fonts API key provided', argParser.usage);
+    }
+    final http.Client client = http.Client();
+    final http.Response response = await client.get(Uri.parse(
+        'https://www.googleapis.com/webfonts/v1/webfonts?key=$apiKey'));
+    if (response.statusCode != 200) {
+      throw ToolExit('Failed to download Google Fonts list.');
+    }
+    final Map<String, dynamic> googleFontsResult = jsonDecode(response.body);
+    final List<dynamic> fontDatas = googleFontsResult['items'] as List<dynamic>;
+    final Map<String, Uri> urlForFamily = <String, Uri>{};
+    for (final Map<String, dynamic> fontData in fontDatas) {
+      if (fallbackFonts.contains(fontData['family'])) {
+        final Uri uri = Uri.parse(fontData['files']['regular'] as String);
+        urlForFamily[fontData['family']] = uri;
+      }
+    }
+    final Map<String, String> charsetForFamily = <String, String>{};
+    final io.Directory tempDir =
+        await io.Directory.systemTemp.createTemp('fonts');
+    for (final String family in fallbackFonts) {
+      print('Downloading $family...');
+      final Uri uri = urlForFamily[family]!;
+      final http.Response fontResponse = await client.get(uri);
+      if (fontResponse.statusCode != 200) {
+        throw ToolExit('Failed to download font for $family');
+      }
+      final io.File fontFile =
+          io.File(path.join(tempDir.path, path.basename(uri.path)));
+      await fontFile.writeAsBytes(fontResponse.bodyBytes);
+      final io.ProcessResult fcQueryResult =
+          await io.Process.run('fc-query', <String>[
+        '--format=%{charset}',
+        '--',
+        fontFile.path,
+      ]);
+      final String encodedCharset = fcQueryResult.stdout;
+      charsetForFamily[family] = encodedCharset;
+    }
+
+    final StringBuffer sb = StringBuffer();
+
+    sb.writeln('// Copyright 2013 The Flutter Authors. All rights reserved.');
+    sb.writeln('// Use of this source code is governed by a BSD-style license '
+        'that can be');
+    sb.writeln('// found in the LICENSE file.');
+    sb.writeln('');
+    sb.writeln('// DO NOT EDIT! This file is generated. See:');
+    sb.writeln('// dev/generate_fallback_font_data.dart');
+    sb.writeln('import \'noto_font.dart\';');
+    sb.writeln('');
+    sb.writeln('const List<NotoFont> fallbackFonts = <NotoFont>[');
+    for (final String family in fallbackFonts) {
+      sb.writeln('  NotoFont(\'$family\', \'${urlForFamily[family]!}\', '
+          '<CodeunitRange>[');
```

Comment:
We should check the code size impact of encoding ranges this way. Perhaps encoding ranges as a flat list of `<int>[start1, end1, start2, end2, ..., startN, endN]` numbers would produce significantly more compact code? I imagine the time difference in initializing the objects would be trivial.

Another thing is, looking at how these ranges are used in `NotoFont.contains` I wonder if we should actually store it as two flat lists of numbers, `rangeStarts` and `rangeEnds`, something like:

```dart
final NotoFont notoSans = NotoFont(
  rangeStarts: <int>[1, 5, 11],
  rangeEnds:   <int>[3, 7, 17],
);
```

Then the `contains` method could binary search `rangeStarts` and `rangeEnds` (should be super fast to binary search flat lists of numbers), and if both searches land on the same range index, then we know that the font contains the code unit.

Reply:
Done